cdap-spark-core-base/src/main/java/co/cask/cdap/app/runtime/spark/classloader/SparkRunnerClassLoader.java

```diff
@@ -107,33 +107,35 @@ public URL apply(String resourceName) {
       return super.loadClass(name, resolve);
     }
 
-    // If the class is already loaded, return it
-    Class<?> cls = findLoadedClass(name);
-    if (cls != null) {
-      return cls;
-    }
-
-    // Define the class with this ClassLoader
-    try (InputStream is = openResource(name.replace('.', '/') + ".class")) {
-      if (is == null) {
-        throw new ClassNotFoundException("Failed to find resource for class " + name);
-      }
-
-      byte[] byteCode = rewriter.rewriteClass(name, is);
-
-      // If no rewrite was performed, just define the class with this classloader by calling findClass.
-      if (byteCode == null) {
-        cls = findClass(name);
-      } else {
-        cls = defineClass(name, byteCode, 0, byteCode.length);
+    synchronized (getClassLoadingLock(name)) {
```

Comment:
is there a performance implication of synchronizing on every class load? Could the findLoadedClass() be outside? Would it help much?

Reply:
It is what's being done in the default java ClassLoader implementation. I don't think that's a big concern here.